src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala

```diff
@@ -25,97 +25,186 @@ import scala.util.{Failure, Try}
+      }
+      resourceTypeId -> roles.toSet
     }
   }
 
-  private def selectActionsForResourceType(resourceTypePK: ResourceTypePK)(implicit session: DBSession): List[ResourceActionRecord] = {
-    val rat = ResourceActionTable.syntax("rat")
-    val actionsQuery =
-      samsql"""select ${rat.result.*}
-               from ${ResourceActionTable as rat}
-               where ${rat.resourceTypeId} = ${resourceTypePK}"""
+  private def loadResourceTypePKsByName(resourceTypes: Iterable[ResourceType])(implicit session: DBSession): Map[ResourceTypeName, ResourceTypePK] = {
+    val rt = ResourceTypeTable.syntax("rt")
+    val query = samsql"""select ${rt.resultAll} from ${ResourceTypeTable as rt} where ${rt.name} in (${resourceTypes.map(_.name)})"""
+    val resourceTypeRecords = query.map(ResourceTypeTable(rt.resultName)).list().apply()
+    resourceTypeRecords.map(rec => rec.name -> rec.id).toMap
+  }
 
-    actionsQuery.map(ResourceActionTable(rat.resultName)).list().apply()
+  override def createResourceType(resourceType: ResourceType, samRequestContext: SamRequestContext): IO[ResourceType] = {
+    upsertResourceTypes(Set(resourceType), samRequestContext).map(_ => resourceType)
   }
 
-  private def selectRolesForResourceType(resourceTypePK: ResourceTypePK)(implicit session: DBSession): List[ResourceRoleRecord] = {
-    val rrt = ResourceRoleTable.syntax("rrt")
-    val actionsQuery =
-      samsql"""select ${rrt.result.*}
-               from ${ResourceRoleTable as rrt}
-               where ${rrt.resourceTypeId} = ${resourceTypePK}
-               and ${rrt.deprecated} = false"""
+  private def upsertRoleActions(resourceTypes: Iterable[ResourceType], resourceTypeNameToPKs: Map[ResourceTypeName, ResourceTypePK])(implicit session: DBSession): Int = {
+    val roleActions = for {
+      rt <- resourceTypes
+      role <- rt.roles
+      action <- role.actions
+    } yield {
+      samsqls"(${resourceTypeNameToPKs(rt.name)}, ${role.roleName}, ${action})"
+    }
+
+    val resourceRole = ResourceRoleTable.syntax("resourceRole")
+    val resourceAction = ResourceActionTable.syntax("resourceAction")
+    val roleAction = RoleActionTable.syntax("roleAction")
 
-    actionsQuery.map(ResourceRoleTable(rrt.resultName)).list().apply()
+    samsql"""delete from ${RoleActionTable as roleAction}
+               using ${ResourceRoleTable as resourceRole}
+               where ${roleAction.resourceRoleId} = ${resourceRole.id}
+               and ${resourceRole.resourceTypeId} in (${resourceTypeNameToPKs.values})"""
+      .update().apply()
+
+    if (roleActions.nonEmpty) {
+      val insertQuery =
+        samsql"""insert into ${RoleActionTable.table}(${RoleActionTable.column.resourceRoleId}, ${RoleActionTable.column.resourceActionId})
+               select ${resourceRole.id}, ${resourceAction.id} from
+               (values $roleActions) AS insertValues (resource_type_id, role, action)
```

Comment:
Can this be changed to use a scalike table like we do with recursive queries?

Reply:
I feel that is overkill. In the recursive queries I like it because it is reused.